### PR TITLE
Fix partial upload bug with invalid remote state

### PR DESCRIPTION
**Pull request by petuhovskiy**
We have an issue that some partial uploaded segments can be actually missing in remote storage. I found this issue when was looking at the logs in staging, and it can be triggered by failed uploads:
1. Code tries to upload `SEG_TERM_LSN_LSN_sk5.partial`, but receives error from S3
2. The failed attempt is saved to `segments` vec
3. After some time, the code tries to upload `SEG_TERM_LSN_LSN_sk5.partial` again
4. This time the upload is successful and code calls `gc()` to delete previous uploads
5. Since new object and old object share the same name, uploaded data gets deleted from remote storage

This commit fixes the issue by patching `gc()` not to delete objects with the same name as currently uploaded.

### CODE REVIEW

**Comment by jcsp** [comment]
Nice catch, how can we write a test reproducer for this case?

**Comment by petuhovskiy** [comment]
> how can we write a test reproducer for this case?

Write some data to db, stop compute, wait for partial upload and simulate S3 errors, then wait for retry with no errors. After that you'll get into a state with missing remote object.

But this test would be too specific, I think it's better to look for these kind of issues in staging/prod. This issue was found because `evict_timeline` validation was failing with ` 1: No file found for the remote object id given` error.

File: safekeeper/src/timeline_eviction.rs [hunk 1]
```diff
@@ -201,7 +201,7 @@ async fn redownload_partial_segment(
     let final_path = local_segment_path(mgr, partial);
     info!(
         "downloaded {} bytes, renaming to {}",
```
Comment (arpad-m) [inline safekeeper/src/timeline_eviction.rs:203]:
```suggestion
        "downloaded {actual_len} bytes, renaming to {final_path}",
```

**Review by arpad-m — APPROVED** [review]
a test would be nice, otherwise looks good to me

**Review by arssher — APPROVED** [review]

**Pull request merged**